### ISSUE TITLE
IP Cortex's WebRTC workshop

### ISSUE THREAD

**Issue by jsms90** [foundersandcoders/master-reference#235]
IP Cortex have suggested that @njsfield and I get involved & potentially create a WebRTC workshop that can be used and replicated for each FAC in the future.

Nick & I are going to record our planning in this issue. @bradreeder @des-des I don't know whether you want us to put a linking document in `fragments`, but we will probably have a repo under `foundersandcoders` to link to :+1:

**Comment by jsms90** [comment]
+ [Beginnings started during FAC9](https://github.com/FAC9/learning-to-use-webrtc)
+ [Will and Marina's successful completion from day 1](https://github.com/FAC9/webrtc-will-marina)

**Comment by jsms90** [comment]
@bradreeder Do you know where we can find the FAC9 s-g-c notes from that week?

**Comment by njsfield** [comment]
Have shared link with FAC9 to share feedback

**Comment by bradreeder** [comment]
@jsms90 @njsfield This is all I found in the s-g-c minutes. Under continue:

* WebRTC workshop challenging and awesome 
(delivery needs improvement)
 (more in tune with the level we are at — intro to promises!)
(introduction too long and hard to follow, nothing visual)
 (smaller groups)

**Comment by bradreeder** [comment]
I have minutes from Marko in my PM I've yet to add. But I can add them to the `london-curriculum` repo. 👍

**Comment by bradreeder** [comment]
@jsms90 @njsfield Are you both happy running the workshop Tues-Thurs? I was planning on reserving Monday for curriculum feedback day (first day of curriculum planning.)

**Comment by njsfield** [comment]
@bradreeder thanks for finding these! Yeah Tues-Thurs should be fine :)

**Comment by peterw8102** [comment]
I'm not sure what lines you're thinking about at the moment but you shouldn't need to write a whole new workshop this time around. The one I did for FAC9 was intended to be replicable across all future workshops, with suitable improvements each time.

All the materials for the workshop, including a fulling working reference implementation, are available in the following github repo:

https://github.com/ipcortex/fac-workshop-materials

**Comment by jsms90** [comment]
@peterw8102 No worries! When Rob came down last week, he briefly mentioned the idea of getting a FAC workshop written, that might be replicable across campuses. But I don't think that was 100% decided at all.

I also didn't realise there was a repo somewhere with everything in one place, including the slides. That's great! I do think there is still a bit of FAC9 feedback that I think could be useful to incorporate before next week though.

Rob also seemed to be talking about @njsfield and I having more of an involvement that Ellie and Elias did last time. But the details weren't clear. If you want Nick and I to just be around for support if/when the students need it, that's fine :+1: It seems like there might just have been a bit of a miscommunication. Whatever you need us for is fine & I'm sure we'll talk about it when you guys head down to London. I'll be available on hangouts :+1:

**Comment by peterw8102** [comment]
@jsms90 you and @njsfield doing as much as possible would be great. I'm very happy taking a back seat - I have rather a lot on my plate at the moment. I think all I'm doing is offering you the workshop and materials from last time as a starting point rather than having to start from scratch. Definitely include feedback from FAC9.

The main changes I made in the repo was to significantly clean up and comment the reference implementation I was using - so you may (or may not!) find that useful.

We can talk more on Thursday though :)

**Comment by jsms90** [comment]
@peterw8102 Sorry, let me be a bit clearer. I'm just hoping that we can add a few things (such as a teeny intro to promises) to make sure that all the groups manage to make progress & no one gets left behind. I really don't want to change it significantly at all :grin: I'm working for you guys because of this workshop after all :wink: 

I'm actually _very_ keen for them to have the same workshop that we did. It was awesome!!! And I definitely wouldn't suggest that you take a back seat. You have so much more experience than Nick and I, and I don't think we should replace you at all!! :upside_down_face: 

Very much looking forward to chatting about next week :tada:

**Comment by jsms90** [comment]
Notes from Thursday:
**Where to start**
FAC9 spent a lot of time setting up an HTTPS server for the first time.
FAC10 have set up a server from scratch more often & have seen HTTPS in week 8 already. They might appreciate more theoretical understanding of how HTTPS works.

Should we start them off with a server already written, to save time? Or would it be beneficial to allow them to do it themselves?
**Action point:** Give boilerplate code. The starting point of the workshop is to clone a repo with an HTTPS server already set up

**Resources before the weekend**
+ Classes
+ Promises

**Gap between high level explanation of problem & setting of task**
+ More intro to WebRTC methods e.g. RTCPeerConnection?

**Comment by njsfield** [comment]
** Day 1 Overview**
1. WebRTC Introduction
2. Discuss plan over 3 days, assert teams of 4.
3. Teams clone HTTPS server
4. Teams serve html page that requests users media and displays webcam in video tag
5. Teams build very simple **local** signalling channel & listener callback function in client side js-
  -  Signalling channel: an object with two methods; **registerUser** and **send**. 
  -  Listener callback: a function with four arguments; **fromEndpoint, toEndpoint, method, data** (for end of day 1, this listener callback function contains simple switch statement which checks if method is 'INIT', if it is, calls console.log with arguments provided)... see [here](https://github.com/FAC9/webrtc-will-marina/blob/master/public/app.js#L130)

**Comment by njsfield** [comment]
I was checking the [MDN Classes](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Classes) article and [MDN Promises](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise) article and would like to recommend them for FAC10 to read before Tuesday

**Comment by jsms90** [comment]
And [MDN on OOP](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Objects/Object-oriented_JS), in case people need some more conceptual explanation as background before the MDN Classes article

**Comment by jsms90** [comment]
I heard that @ZooeyMiller shared a different article on promises that he found more helpful? Could you share this here for posterity please Zooey? :grin:

**Comment by jsms90** [comment]
### FAC9 feedback
[Our SGC](https://github.com/foundersandcoders/london-curriculum/blob/master/stop-go-continue/fac-9/cohort-week-14.md#continue) had this in "Continue":
> WebRTC workshop challenging and awesome  (delivery needs improvement)  (more in tune with the level we are at — intro to promises!) (introduction too long and hard to follow, nothing visual)  (smaller groups)

A couple of weeks before running the workshop, Rob & Pete asked @njsfield and I to help in its delivery. So, iIn the interest of making this an iteratively improving workshop, @njsfield & I gathered some more detailed feedback (specifically looking for constructive cirticisms) from FAC9 in order to make improvements. Here are the results:
+ 'Class' syntax, and 'promises' in IPCortex workshop were not covered in the course, she felt lost at the point they were used in the workshop, and chose not to apply for IPCortex internship as she felt she wouldn't have the skills needed for the work
+ felt totally lost in the workshop, didn't like that there were virtually no resources given to help and would have preferred more guided material (when groups were working alone)
+ Try to set the content/pace so that it is approachable for more than just the most capable ones in the class. It may not have been completely obvious, but least half of FAC9 were no longer engaged by halfway through day 1.
+ Some people came back to it during day 3, but there were 1.5 days where half of the room was not doing the workshop, because they didn't get how.
+ For some people, I think it was quite damaging to their confidence, right before we went out to look for jobs.
+ Don't assume too much knowledge that wasn't in the course
e.g. promises, and even some basics about REST requests (this is largely left to research day during week 3)
+ Big word alerts please!
+ There was quite a jump between the high level explanation during the presentation, and then being able to implement the task. + Even once I understood, at a high level, what I was being asked to do, it felt like I needed to know the RTCPeerConnection API, to have any ideas about where to start. So it seemed like I probably needed to go through the HTML5 Rocks video and/or tutorial before the workshop?
+ It was a bit like asking someone to complete a codewars task before they knew any array methods

**Comment by jsms90** [comment]
Observations
(I'm going to edit this post throughout the 3 days. @njsfield feel free to add to it, so we don't end up with crazy amounts of comments):

Rob's talk
+ A brilliant high level explanation, and breaking this down with very visual diagrams is amazing brilliant! But presentation slides disappear quite quickly, which means that the students are required to take everything in very quickly and remember everything that's said.  
**Action point:** Think about putting the high level explanation diagrams (from Rob's talk) on posters & sticking these around the room.
+ There are still some occasions of "big word alert". The more we notice now, the more we can plan them into the talk for next time. But it's always good to react to the individuals in the room on the fly too (so I did jump in a couple of times):
  + shim
  + polyfill

**Comment by jsms90** [comment]
## Observations
(I'm going to edit this post throughout the 3 days. @njsfield feel free to add to it, so we don't end up with crazy amounts of comments):

### Rob's talk
+ A brilliant high level explanation, and breaking this down with very visual diagrams is amazing! (Something to learn from for API week?) But presentation slides are ephemeral, so it's a shame to have these visual aids disappear quickly, which means that the students are required to take everything in very quickly and remember everything that's said.

**Action point:** Think about putting the high level explanation diagrams (from Rob's talk) on posters & sticking these around the room.

+ There are still some occasions of "big word alert". The more we notice now, the more we can plan them into the talk for next time. But it's always good to react to the individuals in the room on the fly too (so I did jump in a couple of times):
  + shim
  + polyfill

**Comment by ZooeyMiller** [comment]
[Here y'are!](https://scotch.io/@wesleylhandy/writing-your-own-javascript-promises) 😃

**Comment by peterw8102** [comment]
Just reading through these notes from day 1. Feedback from my perspective.

"Big words" - this crops up a lot here. As an external presenter I don't know your curriculum, backgrounds or level of knowledge. I have to make some assumptions about what people know after completing the formal part of the course. Either I assume you know absolutely nothing, in which case nothing would be achieved in the allocated time OR when there is a "big word" that isn't understood someone puts up their hand and say "excuse me, what's an XXX". That really didn't happen here.

> finite state machine (never gave them any material on this)

Yep - I added that over the weekend because it's useful/important for anyone writing software and is actually used in the workshop. I also gave a reference to a short MDN overview of state machines for those interested.

> **Action point**: Go over callbacks & promises slides before splitting into groups. Add the "clone repo" & "get server running" steps on to next bit of workshop

I intentionally changed the order to break out into groups to get the HTTPS server running before going over promises, classes and state machines as a break from dry presentation. By that point people had been sat watching presentation material for around 2 hours, which I think is more than long enough. I was fairly clear that all that was needed was to clone the repo and get the https directory running and showed what they should see. Some groups got there sooner than others and started to look around the rest of the repo, which I think is a good thing.

After the initial presentation the whole of the workshop is going to be breaking out into groups for periods then coming back to the next step via slides. If that's a problem then maybe a better room organisation should be considered?

Gitter channel: that's not something I have access to but I suspect would be useful - or a gitter channel specifically for the 'current workshop' so the person running the workshop can share thing like this with the group. Maybe something set up before the workshop and both students and presenters need to know about beforehand.

> Students were asking what they were supposed to be doing after they cloned the repo and got the server running. **This little bit of setup isn't really something they need to do separately from the rest of it**. They also **don't need to wait for each other to do this before moving on**.
> Nick ended up **posting the instructions in gitter**. It would be better to have this on a slide

Not sure what this means? Didn't need to wait for what? There wasn't a next step to move on to at this point. Putting instructions for what onto gitter? Put what on a slide?

Also probably worth suggesting that if some groups are ahead of others then either they help the others or they take advantage of the time to check out the 'useful reference links' like those on state machines. I think a little initiative can be assumed.

> Explaning REST based polling, HTTP 'Long poll', websockets was much too fast

Fast because you don't at this stage need to know them and some you won't need to know about but the keywords are there if you it sparks interest and you want to go Google them later. If you do need them then they will be covered in more detail later (which I think I said at one point).

**Comment by jsms90** [comment]
Sorry @peterw8102. This wasn't a tactful way to phrase any kind of feedback. All of these notes were taken on the fly, and I wasn't really thinking about how it would come across. I probably shouldn't have put this version online, and instead jotted things down at the time and then posted some more constructive notes at the end.

But in response to your comments
### Big word alerts
> "Big words" - this crops up a lot here. As an external presenter I don't know your curriculum, backgrounds or level of knowledge.

I completely appreciate that. As an external presenter, it is impossible for you to know exactly where the students are at. But this workshop is not a one-off. It is being repeated for multiple cohorts. This is exactly the kind of thing you can use Nick and I to help with. This was not intended as a criticism, but rather a learning point for us to improve the workshop, as it something that is specifically designed to be delivered to FAC students. Ensuring that the language is approachable will have a positive effect in terms of engagement :tada:

("Big word alert" was just a lighthearted reference to a particular video that all the students have watched in their precourse material.)

### Finite state machine
The finite state machine slide was a brilliant addition :heart_eyes:. I just wasn't aware that this would be in the workshop until you were delivering it. These notes were taken in the first day, and the students didn't really touch code that was related to this slide until day 2.

Showing students this slide again on day 2, combined with your idea for students acting this out by passing around pieces of paper was a _perfect_ way to impart this knowledge :+1: 

### When to get students to split into groups
> By that point people had been sat watching presentation material for around 2 hours, which I think is more than long enough

But they split into groups for a very short period of time (just to clone a repo & run a script), before coming back to another hour or so of presentation. I do understand that this long presentation section at the start was because of our feedback about students needing an intro to classes and promises. We could probably impart this info through some more practical exercises next time, to reduce the time before studnets get to start writing code.

> I was fairly clear that all that was needed was to clone the repo and get the https directory running and showed what they should see. Some groups got there sooner than others and started to look around the rest of the repo, which I think is a good thing.

Some groups got there sooner and disengaged from the workshop because they didn't know what they were supposed to be doing. This started a trend throughout the 3 days, where students switched in and out of the workshop, whenever there was a large chunk of time where their group had nothing to do.

> Didn't need to wait for what? There wasn't a next step to move on to at this point.

...exactly.

> After the initial presentation the whole of the workshop is going to be breaking out into groups for periods then coming back to the next step via slides.

We did suggest putting this workshop into a readme instead of slides.

> Also probably worth suggesting that if some groups are ahead of others then either they help the others or they take advantage of the time to check out the 'useful reference links' like those on state machines. I think a little initiative can be assumed.

I did suggest exactly this. Multiple times. But there is value in enabling students to get there for themselves, not hurrying them up & getting them stressed about having to catch up to where the fastest ones are at.

**Comment by jsms90** [comment]
Closing this issue, since it's no longer part of the curriculum
